feat(lidar_centerpoint): enabled per class yaw norm threshold

knzo25 opened the pull request:
Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

## Description
The previous threshold to the yaw norm was applied equally to all classes.
This PR enables this step to use a dedicated threshold per class.

<!-- Write a brief description of this PR. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

yukkysaito commented:
@knzo25 @yukke42 I have one question.
- What is yaw norm?

knzo25 commented:
> @knzo25 @yukke42 I have one question.
> 
> * What is yaw norm?

Centerpoint by default estimates the rotation (yaw) via cosine/sine estimation + atan2.
However, there are no guarantees that the cosine/sine have unit norm (although centerpoint is trained so it does so).
In practice, the detections in which the norm of the cosine/sine estimation is low (e.g., cos=0.000 sin=0.001 -> yaw=90 degrees) are wrong so we filter them out

yukke42 commented on perception/lidar_centerpoint/config/centerpoint.param.yaml:
```diff
@@ -12,3 +12,4 @@
     iou_nms_target_class_names: ["CAR"]
     iou_nms_search_distance_2d: 10.0
     iou_nms_threshold: 0.1
+    yaw_norm_thresholds: [0.5, 0.0, 0.5]
```
From the results of your analysis, I think `0.3` is enough to filter out incorrect orientation estimations. https://tier4.atlassian.net/l/cp/Nqcenf0b

knzo25 replied:
I will update the thresholds right away, but is it ok to put private links here?

knzo25 replied:
*The pointainting config was executable so I removed that permission with `chmod a-x`

yukke42 replied:
> I will update the thresholds right away, but is it ok to put private links here?

Yes. But please care about that the private name isn't included in the link.

knzo25 replied:
:bow:

yukke42 replied:
[1abf696](https://github.com/autowarefoundation/autoware.universe/pull/1962/commits/1abf69663346c6c934228fe22c73bcafb41538a1)

yukke42 commented on perception/lidar_centerpoint/lib/postprocess/postprocess_kernel.cu:
```diff
@@ -124,7 +124,7 @@ cudaError_t PostProcessCUDA::generateDetectedBoxes3D_launch(
     config_.voxel_size_y_, config_.range_min_x_, config_.range_min_y_, config_.down_grid_size_x_,
     config_.down_grid_size_y_, config_.downsample_factor_, config_.class_size_,
-    config_.yaw_norm_threshold_, thrust::raw_pointer_cast(boxes3d_d_.data()));
+    config_.yaw_norm_thresholds_.data(), thrust::raw_pointer_cast(boxes3d_d_.data()));
```
I get an illegal access error.
```
[lidar_centerpoint_node-22] terminate called after throwing an instance of 'thrust::system::system_error'
[lidar_centerpoint_node-22]   what():  reduce failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
```
I think the vector needa to be copied to the device memory to access in the CUDA kernel function.
Here is an example: <https://docs.nvidia.com/cuda/thrust/index.html#iterators-and-static-dispatching>

knzo25 replied:
:bow:

yukke42 replied:
[b730489](https://github.com/autowarefoundation/autoware.universe/pull/1962/commits/b7304898f5a80b2bd381dd774dd67fddbca736de)

yukke42 reviewed (changes requested):
I left a comment.

knzo25 commented:
In addition to addressing your comment, I realized I made a mistake during the review of #1935.
In that PR, `assert(params.target_class_names_.size() == 8);` is no correct, since it contains the names of the classes subject to the algorithm, not the number of total classes.

I deleted said assert in the latest commit, but should I make a separate one-liner PR for that or is it OK to include it here?
Additionally, instead of `constexpr std::size_t num_object_classification = 8;`, shouldn't it be better to add a constant in the message file to make is usable everywhere?

yukke42 commented:
> I deleted said assert in the latest commit, but should I make a separate one-liner PR for that or is it OK to include it here?

Thanks! Yes, It's ok. 

> Additionally, instead of constexpr std::size_t num_object_classification = 8;, shouldn't it be better to add a constant in the message file to make is usable everywhere?

I think we should add the constant value to the perception_utils package, but let's do it after <https://github.com/autowarefoundation/autoware.universe/pull/1954>  since this is a big refactoring.

yukke42 commented on perception/lidar_centerpoint/lib/postprocess/postprocess_kernel.cu:
```diff
@@ -116,6 +117,9 @@ cudaError_t PostProcessCUDA::generateDetectedBoxes3D_launch(
   const float * out_rot, const float * out_vel, std::vector<Box3D> & det_boxes3d,
   cudaStream_t stream)
 {
+  thrust::device_vector<float> yaw_norm_thresholds(
+    config_.yaw_norm_thresholds_.begin(), config_.yaw_norm_thresholds_.end());
```
I think it's better to declare `yaw_norm_thresholds` as as a class member since memcopy at once is enough. And in this package I use `_d` as a notation meaning the device_vector / device_pointer.

knzo25 replied:
Done!

yukke42 replied:
[3288912](https://github.com/autowarefoundation/autoware.universe/pull/1962/commits/3288912a621b137787006ca91fdaa682d1a2d107)

yukke42 reviewed (approved):
LGTM. I check that there is no error.
